### DISCUSSION TITLE
Relaunching Labels! 🎉

### DISCUSSION THREAD

**Discussion by andrewhong5297** [duneanalytics/spellbook#2634] [Sector: Labels]
# Relaunching Labels

Labels were first launched in October 2022, and have seen a steady growth from individual contributors in both static (manually labeled addresses) and query-based sources. Here is the current schema:

![Untitled](https://user-images.githubusercontent.com/47720952/216987836-5ab3e891-56c8-4994-832f-52e1ab82f99c.png)

We want to make it easier to contribute and use labels, as right now, there is very little structure for query labels (those outside of static labels like CEX or multisig). 

The first step towards doing this is being more specific on what can go into a `category`, and adding a `label_type` column that clarifies the types of labels we will accept. There will also be a `label_name` column to make it easier to navigate aggregated labels tables by category. There will also be a `version` column that defaults to 1 for all labels until there is some conflict (i.e. two different methodologies for the same model.

### Label Types:

Here’s a list of label types:

- **Identifiers**: Most static labels should be this label type, as well as common usernames such as Farcaster, ENS, and Lens names. As a rule of thumb, identifiers should usually specify a unique entity name.
- **Usage:** These are the existing top volume and frequency (or some other percentile-able metric) within a domain and the usage of specific protocols. There must be some sort of ranking/percentile involved!
- **Personas:** These are for on-chain curated behaviors (like common CT memes) or protocol user tagging. They should be easily understood to non-analysts, though the underlying calculation methods may be more subjective. 

This is just the starting list. If a label is created that clearly fits inside a new type, we will create a type for it. The idea is to try and keep types smaller in number and very clear so that it's easier for analysts to navigate, use, and contribute to.

## Example Hierarchies:

To make this clear, let’s do some examples.

**Category: NFT**

- Label_types
    - Identifier: Opensea Username (traderName_opensea)
    - Usage: top volume by nft.trades
    - Usage: top transactions by nft.trades
    - Persona: Opensea User, Sudoswap User, Blur User
    - Persona: Wash Trader
    - Persona: Art Blocks Curator (holds all art blocks curated collections)
    - Persona: Airdrop Hunter
    - Persona: Early NFT Trader (first year of nft trading)

**Category: Social**

- Label_types
    - Identifier: Lens username (.lens) ENS reverse resolver (.eth), farcaster (_farcaster)
    - Usage: top holders from ENS
    - Usage: top posters from lens
    - Persona: Lens User, ENS User
    - Persona: Squatter (sitting on dozens of ENS names)

**Category: DEX**

- Label_types
    - Usage: top volume by dex.trades
    - Usage: top transactions by dex.trades
    - Persona: Uniswap v2 User, Cowswap User
    - Persona: Sandwiched User
    - Persona: Yield Farmer
    - Persona: Liquidity Provider

### Migration of Current Labels:

I promise it isn’t as scary as it sounds. We will start by reorganizing the folder of [labels](https://github.com/duneanalytics/spellbook/tree/main/models/labels/dex_traders) structure to look like this:

```python
"""
Labels (sector)
- NFT (category)
    - label_nft.sql ← aggregates all labels from the folders below
    - Identifier (folder)
        - label_opensea.sql
    - Usage (folder)
	- Ethereum (folder)
	- label_nft_volume_percentile.sql
    - Persona (folder)
	- Ethereum (folder)
	- label_wash_trader_millions.sql
"""
```

At a glance, it should be easy to explore from category down to label level.

After the folders and files are reorganized, we can clean up the actual schemas/sql. Most static labels need a `label_type` column added, and then that’s it. Many of the query-based models like `dex_traders` ([here](https://github.com/duneanalytics/spellbook/tree/main/models/labels/dex_traders) will need to be changed; henrystats and gentrexha mainly created these. 

Models like `ens` and `contracts` will be hard to edit without breaking many people’s queries, so we will leave that alone.

For aggregated labels, there will also be a `label_name` column added. so within `label_nft.sql` , the `label_name` for `label_nft_volume` will be `nft_volume_percentile`.

⚠️ **Please reply with any of your concerns/comments; once we have a general agreement, we can start work and codify this into a labels readme.**

Once the structure of labels is in a good state and all cleaned up, we can launch a hackathon with rewards for the most creative, useful, memetic, etc. We’re really excited about unlocking the potential of labels more (and making it more multiplayer)!

**Comment by andrewhong5297** [comment]
tagging in all contributors of labels to kickoff the discussion! If you're okay with the changes, please give a thumbs up. When we have approval from enough contributors then I will go ahead and refactor the folders/labels (everything besides `ens` and `contracts`). 

You all will be tagged on the PR again for final review before it goes through.

@hildobby @gentrexha @henrystats @MSilb7 @markusbkoch @Hosuke @Stefenon @0xRobin @agaperste @jeff-dude @soispoke @catherine-2 @AlexTheLion123 @jackiep00

**Reply by hildobby** [reply]
I'm onboard for standardisation, I agree that it's time to clarify what is meant by each label and some more standardisation/rules will help a lot in that regard.

Maybe we can also have a place to write a quick few sentences about what each label category means and have that required for new labels? Most would be extremly obvious but it can pretty complex imo so probably worth it 🤷‍♂️

**Reply by agaperste** [reply]
I like the idea - have short sentence descriptions. Feels like we can enforce that when PR is created for a new label category, i.e. schema yaml?

**Reply by andrewhong5297** [reply]
@markusbkoch @gentrexha Please see this draft PR to better understand how folders/models will be organized under the refactor. 

https://github.com/duneanalytics/spellbook/pull/2659/files

I've only done DEX labels so far to keep it simpler. I will be moving around other labels here to consolidate categories after we have agreement.

**Reply by andrewhong5297** [reply]
we could also add an aggregate view at the usage, persona, and identifier levels to then make the overall labels_dex aggregation easier.

**Reply by andrewhong5297** [reply]
Okay, the final PR is up for review. You can [explore around](https://github.com/duneanalytics/spellbook/tree/labels-relaunch-2023-final/models/labels) what it'll feel like and give feedback [in the new PR](https://github.com/duneanalytics/spellbook/pull/2680) until end of day 2/16 (Thursday).

**Reply by andrewhong5297** [reply]
Reminder last day to add any comments! :)

**Comment by dot2dotseurat** [comment]
For `personas` and `usage` types, I expect there are multiple ways to calculate or think about them. Do we want to have a framework where one definition is accepted or do we want to be able to have competing versions? I can think of scenarios where we might have uploaded the results of an ML-based model for identifying wash traders versus a strict SQL logic version etc. 

The trade-off is that it's more confusing when there are multiple versions, but it's also a more flexible framework.

**Reply by andrewhong5297** [reply]
Great question! Maybe a good way to clarify is that "Usage" labels have to have some sort of ranking involved? like top 10% wash trader. Versus "Persona" will not (so just "Wash Trader").

**Reply by andrewhong5297** [reply]
I've updated the post to be more clear on this, thank you.

**Reply by dot2dotseurat** [reply]
Right but what if there are still competing methodologies to calculate top 10%. E.g. some people think a certain type of trade should be excluded from the ranking etc. How do we want to govern the definitions?

**Reply by andrewhong5297** [reply]
Ah I misunderstood. I think it will be exciting if we end up with a disagreement on a label, and we can host a public discussion about it and then a poll for contributors. 

For conflicts like the one you proposed, labels just inherit from wash trades so if some person has an issue with what trades are being included then that is a wash trades table discussion, not a label one. The benefit of building more labels off of spells is the inherited data quality and aligned understanding. We should generally push for labels to be built off of existing spells more than raw/decoded data as much as we can (to minimize debates that can happen at a lower feed level).

**Reply by dot2dotseurat** [reply]
Ok, my rec is just that we have a `version` column so we are prepared for instances where the resolution to the disagreement on a label methodology is we have multiple versions that won't result in a schema change. Does that seem reasonable?

**Reply by hildobby** [reply]
Yeah I do expect some conflicting ideas behind different models and we could very well see multiple version of the same label but then wouldn't the creator be a sufficient differentiator? Adding a `version` column could also work ofc

**Reply by andrewhong5297** [reply]
Added version column that defaults to 1!

**Reply by andrewhong5297** [reply]
Maybe in the future there will be some "methodology" column to add, but can wait and see what happens :)

**Comment by markusbkoch** [comment]
Sorry if none of what I'm going to say makes sense. I'm not familiar with labels on Dune V2, only [V1](https://github.com/duneanalytics/spellbook/pull/171).

Labels are extremely useful and we've grown to become quite dependent on them in our queries and dashboards, and have been missing them since https://dune.com/labels was taken down. 

IIUC the proposed architecture reduces creator's freedom by limiting them to a fixed set of label types, whereas today (or in V1 at least) they can/could create types and names. 

My concern is that this risks stifling innovation. To illustrate what I mean, imagine we are labelling animals instead of addresses and we define that the only label types allowed are `genus` and `species`. Now if someone's use case requires that they label an animal according to their `family` or `predominant_color` or `cuteness_score`, they can't.

In other words, users should continue to be able to create dimensions (types) and assign to objects (addresses) a value (name) in that dimension.

**Reply by augustog** [reply]
This is a very good point @markusbkoch! I don't see a reason why we wouldn't add categories and label types when needed in the future, so if anyone needs to extend the taxonomy to accommodate a new use case I'd say we should have space for that. 

@andrewhong5297 wdyt?

**Reply by markusbkoch** [reply]
I'm sure the Dune team would be open to adding new categories and labels, but there's a big difference between having to go through that process and being able to permissionlessly and immediately create a type like we used to be able to in V1.

**Reply by markusbkoch** [reply]
If users lose that ability, I'm pretty sure they'd hack their way into getting it back by baking their custom types within the label name (eg. type=`identifier`, name=`color-black`). Sure, the performance would be terrible, but it would bring back the V1 experience (which from that user's perpective is just feature completeness, really)

**Reply by dot2dotseurat** [reply]
I don't think the permission should come from the Dune team. But I do think V1 labels ended up being a bit of a mess without any type of governance. My general preference would be something like adding a new category needs some level of review from active contributors. Otherwise, I think it would be easy for categories to become somewhat chaotic.

**Reply by andrewhong5297** [reply]
great points to everyone here. 

So `label_type` is meant to tag the type of label for easy filtering. You can still call your label whatever you want in the model name. That model name then gets put into part of the aggregated schema so `label_wash_traders.sql` as a persona model would be called `wash_traders` at the `labels_nft.sql` category aggregated model. 

In your example, the model would be called `label_color.sql` with the `name` being "black", `label_type` being "identifier". In the aggregated category file the `label_name` would be "color". I think maybe having "name" used in two columns is confusing. The main point of this is to keep `category` and `label_type` as clean as possible, so you don't need to some weird LIKE filtering to get all DEX labels.

Let me know if that's still unclear, I'll try to clarify it in the proposal above.

**Reply by andrewhong5297** [reply]
Also want to add that we're happy to add label types if they really don't fit an existing category. Just figured that all our current labels fit into these three types.

**Reply by 0xRobin** [reply]
> In other words, users should continue to be able to create dimensions (types) and assign to objects (addresses) a value (name) in that dimension. -- @markusbkoch 

So from @andrewhong5297's comments this can still be achieved where the model name (label_name) can be whatever you want to call your dimensions?

+1 btw on not having 2 `name` columns. In my head a label has a name (ENS) and a value (0xrob.eth) and belongs to an address (0x5f8..) on a chain (ethereum), hope that made sense.

**Reply by markusbkoch** [reply]
>So from @andrewhong5297's comments this can still be achieved where the model name (label_name) can be whatever you want to call your dimensions?

That was my first impression after reading Andrew's latest comments. But I'm not 100% sure, since these references to "aggregation" confuse me:
> In the aggregated category file the label_name would be "color".
> For aggregated labels, there will also be a label_name column added

+1 to this:
> In my head a label has a name (ENS) and a value (0xrob.eth) and belongs to an address (0x5f8..) on a chain (ethereum), hope that made sense.

**Reply by markusbkoch** [reply]
> I do think V1 labels ended up being a bit of a mess without any type of governance

@dot2dotseurat can you elaborate on this? How did you use V1 labels and how was that messy? I've always felt that being able to filter labels by author+type gave me all I needed to separate the signal from the noise. I can choose which ones to trust for each purpose. Here's [one example](https://dune.com/queries/26390/53561), where we define a set of Dune users that we trust to properly label Balancer LBPs so that we could use those labels as "friendly names" in dashboards.

**Reply by andrewhong5297** [reply]
I'll start a PR with a few models just to give a better idea, think it is hard for me to just explain in text 😅

**Reply by dot2dotseurat** [reply]
@markusbkoch sure, when I started looking at Dune labels I was new to the platform so I didn't have the context of which authors/type to trust. I think that relying on trust for a certain author works if you have a ton of context but gets tricky otherwise.

**Reply by andrewhong5297** [reply]
I've replied to an earlier comment with a PR example. I think in the example given, we would expect to just be a generated view or a model in the balancer sector in spellbook instead of in labels.

Being more intentional here will keep labels from bloating too much.

**Reply by markusbkoch** [reply]
Thank you for the PR. [This](https://github.com/duneanalytics/spellbook/pull/2659/commits/c4fa87849acfa07e583a3443d18c5afa5cab2bfa#diff-37d3a043f5d6655ff56e20c0cbe4274914c243f74ecafa3fd4c59be9aa1d5141R33) and [this](https://github.com/duneanalytics/spellbook/pull/2659/commits/c4fa87849acfa07e583a3443d18c5afa5cab2bfa#diff-72986296362690219349b9bed0e240d787d7de30ee5406d31f2b6dc8b75e9813R43) for example seem to confirm my and 0xRobin's previous understanding that `model_name` is the equivalent of V1's `label.type`, which grants the model creator the same level of flexibility that they used to have. Which is good.

>I think in the example given, we would expect to just be a generated view or a model in the balancer sector in spellbook instead of in labels.

I can understand where you are coming from, but this model would not have worked for our use case at the time. 

LBPs are ephemeral, they don't usually last more than a few days. We needed the dashboard to be updated as soon as a new LBP was launched. Every piece of onchain data was immediately updated automatically, because all these pools come from factories so we don't have to submit contracts for decoding every time. But the missing piece, the "friendly name" lived off-chain. So we counted on the biz-dev team to contribute to the labels dataset via https://dune.com/labels.

https://dune.com/labels was super powerful because it was the first time we could merge on-chain and off-chain data on Dune. Technical and non-technical users could cooperate to develop something bigger that either group could have achieved by themselves. Seamlessly and instantly - no need to wait for a PR to be reviewed and merged!

I'll add another example. We have dashboards to help our community understand what drives volume on Balancer. `dex.trades` tells us what contract a user originally interacted with that caused a swap to occur. But we need to aggregate all the different 1inch router addresses into a single entity. We use [labels](https://dune.com/queries/22281/46099) for that. We are currently only using labels `where "author" = 'balancerlabs'`, so one might think that your proposed view/model in the balancer sector would work just as well. But it prevents us from:
1. responding quickly to requests for enriching the dashboard; in V1 all we had to do was go to https://dune.com/labels, now we'd have to submit a PR and wait for review
2. evolving to a final, more efficient state, where we crowdsource this categorization to trustworthy members of the Dune/Balancer communities.

**Reply by andrewhong5297** [reply]
I totally see your point. I think more easily updatable static labels are an important problem to solve, following this labels refactor. 

Maybe worth brainstorming with @augustog and @dot2dotseurat on how users could add quickly to static labels - maybe we can leverage the query-a-query feature such that part of the labels model could pull from query_1234 as a source so that balancer could update static labels in a query only they control and without our approval, and it would propogate into labels in spellbook automatically.

Would be a nice solve for project owned static labels, but maybe there is a better design.

**Reply by 0xRobin** [reply]
Chiming in that as a generic workaround for having any static "personal" (not used by other parties) data in Dune.
You can have a static values query in DuneSQL that contains all your off-chain information in a view, and then leverage this view in your other queries.
https://dune.com/docs/reference/dune-v2/query-engine/#query-queries-as-views-in-dune-sql
This allows you to have direct control over the data (and no wait times for PR approvals)

**Comment by gentrexha** [comment]
I see that you, in the current schema, mention that the blockchain column is of type string. This wasn't the case until now, and it was of type List[string]. 

Is the plan to use just string from now? I've never understood the rationale behind the original type.

**Reply by andrewhong5297** [reply]
I think it would be easier to just use a string, and maybe have a "cross-chain" string for labels that apply to multiple chains. 

This does affect current queries though, so something to keep in mind.

**Comment by andrewhong5297** [comment]
this is now closed, the labels have been merged :) you can query explore them here https://dune.com/queries/2010522?d=11&category_e15077=dex
